sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -246,6 +247,54 @@ class Analyzer(
       CleanupAliases)
   )
 
+  /**
+   * 1. Turns Add/Subtract of DateType/TimestampType/StringType and CalendarIntervalType
```

Comment:
I think we need to carefully design the rule so that it's easy to reason about. How about
```
For `UnresolvedAdd`:
1. if one side is interval, turn it into `TimeAdd`
2. if one side is date, turn it into `DateAdd`
3. otherwise, turn it into `Add`

For `UnresolvedSubtract`:
1. if the right side is interval, turn it into `TimeSub`
2. if one side is timestamp, turn in into `SubtractTimestamps`
3. if one side is date, turn it into `DateDiff`/`SubtractDates`
4. otherwise, turn it into `Subtract`
```

Reply:
Update a bit
```scala
  /**
   * For [[UnresolvedAdd]]:
   * 1. If one side is timestamp/date/string and the other side is interval, turns it to
   * [[TimeAdd]];
   * 2. else if one side is date, turns it to [[DateAdd]] ;
   * 3. else turns it to [[Add]].
   *
   * For [[UnresolvedSubtract]]:
   * 1. If the left side is timestamp/date/string and the right side is an interval, turns it to
   * [[TimeSub]];
   * 2. else if one side is timestamp and the other side is date/timestamp, turns it to
   * [[SubtractTimestamps]];
   * 3. else if both side are dates, turns it to [[DateDiff]]/[[SubtractDates]];
   * 4. else if the left side is date, turns it to [[DateSub]];
   * 5. else turns it to [[Subtract]].
   *
   * For [[UnresolvedMultiply]]:
   * 1. If one side is interval and the other side is numeric, turns it to [[MultiplyInterval]];
   * 2. otherwise, turns it to [[Multiply]].
   *
   * For [[UnresolvedDivide]]:
   * 1. If the left side is interval and the right side is numeric, turns it to
   * [[DivideInterval]];
   * 2. otherwise, turns it to [[Divide]].
   */
```